src/com/sun/jna/Memory.java

```diff
@@ -179,15 +188,31 @@ public Memory align(int byteBoundary) {
     /** Properly dispose of native memory when this object is GC'd. */
     @Override
     protected void finalize() {
-        dispose();
+        // if the dispose call is done by the Finalizer the key is not accessible in a WeakHashMap any more
+        // the call of allocatedMemory.remove is a waste of cpu time in this case
+        dispose0(false);
     }
 
     /** Free the native memory and set peer to zero */
-    protected synchronized void dispose() {
+    protected void dispose() {
+        dispose0(true);
+    }
+
+    /** Used to handle different types of dispose calls.
+     * @param maintainMap whether the {@code allocatedMemory} map needs to be updated
+     */
+    private synchronized void dispose0(boolean maintainMap) {
         try {
             free(peer);
         } finally {
-            allocatedMemory.remove(this);
+            synchronized (allocatedMemory) {
+                if (maintainMap) {
+                    allocatedMemory.remove(this);
+                } else {
+                    // ensures the stale map entries are removed from the map
+                    allocatedMemory.size();
```

Comment:
Is this a significant performance improvement? You are already paying the price for the synchronized block, why not unconditionally call `allocatedMemory#remove`?

If this is not measureable, I would prefer to fold this into dispose and drop the `maintainMap` parameter.

Reply:
This synchronized block has been done by the Collections.synchronizedMap wrapper before, it only has moved. 
It's not that easy to mearsure the improvement, because it's done by the Finalizer Thread and there might be no impact to the thread using the Memory instances.
I think it might be possible to replace the Map by a implementation like the ProxyObject does… (see the comment below regarding the next steps).